File: 1006-vowel-spellchecker/vowel-spellchecker.py
import re
from collections import defaultdict
from typing import List

class Solution:
    def spellchecker(self, wordlist: List[str], queries: List[str]) -> List[str]:
        def devowel(word):
            return re.sub(r'[aeiou]', '*', word.lower())
        
        # Sets and dictionaries
        original_words = set(wordlist)
        lowercase_map = dict()
        vowel_map = dict()
        
        # Store first occurrence only for maps
        for word in wordlist:
            lower = word.lower()
            mask = devowel(word)
            if lower not in lowercase_map:
                lowercase_map[lower] = word
            if mask not in vowel_map:
                vowel_map[mask] = word
        
        result = []
        for query in queries:
            if query in original_words:  # Rule 1
                result.append(query)
                continue
            lower_query = query.lower()
            if lower_query in lowercase_map:  # Rule 2
                result.append(lowercase_map[lower_query])
                continue
            mask_query = devowel(query)  # Rule 3
            if mask_query in vowel_map:
                result.append(vowel_map[mask_query])
                continue
            result.append("")  # No match
        return result

File: 1006-vowel-spellchecker/vowel-spellchecker.java
import java.util.*;

class Solution {
    private static final String VOWELS = "aeiou";

    public String[] spellchecker(String[] wordlist, String[] queries) {
        Set<String> words = new HashSet<>(Arrays.asList(wordlist));
        Map<String, String> caseInsensitive = new HashMap<>();
        Map<String, String> vowelMasked = new HashMap<>();

        for (String word : wordlist) {
            String lower = word.toLowerCase();
            caseInsensitive.putIfAbsent(lower, word);
            String mask = devowel(lower);
            vowelMasked.putIfAbsent(mask, word);
        }

        String[] result = new String[queries.length];
        for (int i = 0; i < queries.length; i++) {
            String query = queries[i];
            if (words.contains(query)) { // Exact match
                result[i] = query;
            } else {
                String lower = query.toLowerCase();
                if (caseInsensitive.containsKey(lower)) { // Case-insensitive match
                    result[i] = caseInsensitive.get(lower);
                } else {
                    String mask = devowel(lower);
                    if (vowelMasked.containsKey(mask)) { // Vowel error match
                        result[i] = vowelMasked.get(mask);
                    } else {
                        result[i] = ""; // No match
                    }
                }
            }
        }
        return result;
    }

    private String devowel(String word) {
        StringBuilder sb = new StringBuilder();
        for (char c : word.toCharArray()) {
            if (VOWELS.indexOf(c) >= 0)
                sb.append('*');
            else
                sb.append(c);
        }
        return sb.toString();
    }
}
